src/graphics/material/new_material.rs

```diff
@@ -0,0 +1,247 @@
+//Materials are being Stored in a HashSet
+use super::Image;
+use crate::graphics::resources::BindGroup;
+use bytemuck::{Pod, Zeroable};
+use nalgebra_glm::{vec4, Vec4};
+use serde;
+use std::{hash::Hash, mem, sync::Arc, collections::HashMap};
+use walkdir::WalkDir;
+
+#[derive(PartialEq)]
+pub enum MaterialKind {
+    Unlit,
+    PBR,
+    None,
+}
+impl From<&NewMaterialHandle> for MaterialKind {
+    fn from(h: &NewMaterialHandle) -> Self {
+        if h.main_texture.is_some()
+            && h.roughness_texture.is_some()
+            && h.normal_texture.is_some()
+            && h.roughness.is_some()
+            && h.metallic.is_some()
+        {
+            MaterialKind::PBR
+        } else if h.main_texture.is_some() && h.color.is_some() {
+            MaterialKind::Unlit
+        } else {
+            MaterialKind::None
+        }
+    }
+}
+/// Hash as identifier.
+pub struct NewMaterialData {
+    pub material_kind: MaterialKind,
+    pub main_texture: Option<Arc<Image>>,
+    pub roughness_texture: Option<Arc<Image>>,
+    pub normal_texture: Option<Arc<Image>>,
+    pub roughness: Option<f32>,
+    pub metallic: Option<f32>,
+    pub color: Option<[f32; 4]>,
+    pub uniform_buf: Option<wgpu::Buffer>,
+}
+
+impl NewMaterialData {
+    pub(crate) fn create_bind_group<'a>(
```

Comment:
I'm not sure this is the best place for this to live now. As it's very dependent on which pipeline is being used. For example the unlit shader doesn't care about `PBRMaterialUniform` at all. We could make the data consistent across shaders, and not care that we are sending a little extra information alongside?

Reply:
Yea I wasnt sure about this. I was trying to only have one material for everything. We would have Optional fields for every possibility and the MaterialKind would specify which shader and bind group is needed for it